### PR TITLE
Astro LP price source update to match alphafinance impl

### PR DESCRIPTION
**Pull request by piobab**
https://blog.alphaventuredao.io/fair-lp-token-pricing/

### CODE REVIEW

File: contracts/mars-oracle/src/contract.rs [hunk 1]
```diff
@@ -285,8 +286,8 @@ fn query_asset_price(
             Ok(price)
         }
 
-        // The value of each unit of the liquidity token is the total value of pool's two assets
-        // divided by the liquidity token's total supply
+        // The calculation of the value of liquidity token, see: https://blog.alphafinance.io/fair-lp-token-pricing/.
+        // This formulation avoids a potential sandwich attack that distorts asset prices by a flashloan.
```
Comment (larry0x) [inline contracts/mars-oracle/src/contract.rs:290]:
```suggestion
        // This formulation avoids a potential manipulation attack that distorts asset prices by a flashloan.
```

File: contracts/mars-oracle/src/contract.rs [hunk 2]
```diff
@@ -1162,6 +1197,149 @@ mod tests {
...
+                    deps.as_ref(),
+                    mock_env(),
+                    QueryMsg::AssetPriceByReference {
+                        asset_reference: asset_reference.clone(),
+                    },
+                )
+                .unwrap(),
+            )
+            .unwrap();
+            assert_eq!(price, Decimal::from_ratio(211_u128, 1000_u128));
```
Comment (larry0x) [inline contracts/mars-oracle/src/contract.rs:1300]: can you write some comments on how to arrive at the price? example: https://github.com/mars-protocol/fields-of-mars/blob/v1.0.0/scripts/tests/4_martian_field.test.ts#L418-L421

Comment (larry0x) [reply contracts/mars-oracle/src/contract.rs:1300]: it'd also be helpful to have a unit test to simulate the case where the price is manipulated (e.g. someone dumps a huge amount of Luna in a previous tx of the same block) and see if the calculated price is affected

**Comment by larry0x** [comment]
@piobab LGTM, with a few comments

**Review by dancreee — APPROVED** [review]

File: contracts/mars-oracle/src/contract.rs (hunk 2 above)
Comment (piobab) [reply contracts/mars-oracle/src/contract.rs:1300]: @larry0x could you please develop your thought? The `AstroportLiquidityToken` price logic takes only pool response from Astro (which is mocked and doesn't have relation to other tx, block etc so can't be simulated in the way you wrote) and asset prices from other price sources (which are already covered with tests). Maybe I didn't fully understand you.

Comment (larry0x) [reply contracts/mars-oracle/src/contract.rs:1300]: @piobab don't worry i'll add a commit to this PR. just want to add an extra test case.

**Review by alienHunterOnMars — APPROVED** [review]

File: contracts/mars-oracle/src/contract.rs (hunk 2 above)
Comment (larry0x) [reply contracts/mars-oracle/src/contract.rs]: @piobab done: https://github.com/mars-protocol/mars-core/pull/323/commits/a0a9c5fca80930e980c0668b65d6fe69cebce52a

**Comment by larry0x** [comment]
@dancreee @alienHunterOnMars i added an extra unit test to the PR. can you approve again

**Review by larry0x — DISMISSED** [review]

File: contracts/mars-oracle/src/contract.rs [hunk 3]
```diff
@@ -1162,6 +1197,198 @@ mod tests {
...
+                .unwrap(),
+            )
+            .unwrap();
+
+            // Assume we use Terra L1 oracle for Luna and UST, their prices should not be affected
+            // LUNA price: 88.5
+            // UST price: 1
+            // pool value = 2 * sqrt((88.5 * 6389) * (1 * 1385000)) = 1769874
+            //
+            // Is slightly (<0.01%) off from the pre-manipulation value. Good enoguh!
```
Comment (piobab) [inline contracts/mars-oracle/src/contract.rs:1348]:
```suggestion
            // Is slightly (<0.01%) off from the pre-manipulation value. Good enough!
```

**Review by larry0x — APPROVED** [review]

**Review by dancreee — APPROVED** [review]

**Comment by larry0x** [comment]
one thing that just came to mind is that when creating the price source, we should assert that the pool is an XYK pool, not a stableswap pool. i'm gonna do a commit for this if it's ok with you @piobab

**Comment by larry0x** [comment]
> one thing that just came to mind is that when creating the price source, we should assert that the pool is an XYK pool, not a stableswap pool. i'm gonna do a commit for this if it's ok with you @piobab

commit pushed, pls review

**Comment by piobab** [comment]
> > one thing that just came to mind is that when creating the price source, we should assert that the pool is an XYK pool, not a stableswap pool. i'm gonna do a commit for this if it's ok with you @piobab
> 
> commit pushed, pls review

LGTM

**Review by larry0x — APPROVED** [review]

**Review by dancreee — CHANGES_REQUESTED** [review]
@piobab LGTM just need to resolve merge conflicts from #320

**Review by dancreee — APPROVED** [review]

**Pull request closed without merging**